app_flutter/lib/service/appengine_cocoon.dart

```diff
@@ -100,6 +101,47 @@ class AppEngineCocoonService implements CocoonService {
     return response.statusCode == HttpStatus.ok;
   }
 
+  /// Downloads the log for [task] and writes the bytes to the local storage of
+  /// the current device. Returns true if write was successful, and false if
+  /// there was a failure. Only works on the web platform.
+  ///
+  /// Steps to download the log:
+  /// 1. Make an authenticated request to the Cocoon backend for the task log.
+  /// 2. Encode the log to be uri safe.
+  /// 3. Create an anchor element with the encoded log data for downloading.
+  /// 4. Click the anchor element to trigger the browser to download the log.
+  @override
+  Future<bool> downloadLog(Task task, String idToken, String commitSha) async {
+    assert(task != null);
+    assert(idToken != null);
+
+    /// This is a web only solution for downloading logs.
+    if (!kIsWeb) {
+      return false;
+    }
+
+    final String getTaskLogUrl =
+        _apiEndpoint('/api/get-log?ownerKey=${task.key.child.name}');
+
+    // This line is dangerous as it fails silently. Be careful.
+    html.document.cookie = 'X-Flutter-IdToken=$idToken;path=/';
+    // This wait is a hack as the above line is not synchronous. It takes time
+    // to write the cookie back to the browser. This is required before making
+    // the download request.
+    await Future<void>.delayed(const Duration(milliseconds: 500));
```

Comment:
Any way to avoid the hack?

Reply:
One design I have in mind is to poll the cookies for some duration until the cookie has been written. This polling would wait until it reads the current id token for `X-Flutter-IdToken` or fail on some timeout. However, this design assumes that the `dart:html` interface for reading reads from the browser instead of what dart stores in memory.

Reply:
I need more research on my part to see if the above is a feasible approach.